### PR TITLE
[storage] get_last_version_before_timestamp()

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation
Via binary search on the event by key index.
This is for audit report generation. There's on proof in this version, but proof can be two consecutive NewBlockEvents (with their proofs) showing the timestamp bump from before to at/after the timestamp in the request.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan
Unit test.

### CODE REVIEW

File: storage/diemdb/src/event_store/mod.rs [hunk 1]
```diff
@@ -182,6 +207,25 @@ impl EventStore {
         Ok(result)
     }
 
+    fn lookup_event_by_key(
+        &self,
+        event_key: &EventKey,
+        seq_num: u64,
+        ledger_version: Version,
+    ) -> Result<(Version, u64)> {
+        let indices = self.lookup_events_by_key(&event_key, seq_num, 1, ledger_version)?;
```
Comment (kchalkias) [inline storage/diemdb/src/event_store/mod.rs:216]: do you need the `&` to `event_key`? it's already a reference.

File: storage/diemdb/src/event_store/mod.rs [hunk 2]
```diff
@@ -238,6 +282,80 @@ impl EventStore {
...
+        mut comp: C,
+        ledger_version: Version,
+    ) -> Result<Option<u64>>
+    where
+        C: FnMut(&ContractEvent) -> Result<bool>,
+    {
+        let mut begin = 0u64;
+        let mut end = match self.get_latest_sequence_number(ledger_version, event_key)? {
+            Some(s) => s
+                .checked_add(1)
```
Comment (kchalkias) [inline storage/diemdb/src/event_store/mod.rs:300]: why do we need this?

File: storage/diemdb/src/event_store/mod.rs [hunk 3]
```diff
@@ -238,6 +282,80 @@ impl EventStore {
...
+                .ok_or_else(|| format_err!("event sequence number overflew."))?,
+            None => return Ok(None),
+        };
+
+        let mut count = end - begin;
```
Comment (kchalkias) [inline storage/diemdb/src/event_store/mod.rs:305]: Why using a temp `counter` and not the common way of doing binary search where we check `while (begin <= end)`? I.e., as in https://dev-notes.eu/2020/03/Binary-Search-in-Rust/

Comment (msmouse) [reply storage/diemdb/src/event_store/mod.rs:305]: http://www.cplusplus.com/reference/algorithm/lower_bound/ 
https://github.com/microsoft/STL/blob/b52c3797e262cbce28166d92f9d1b8f2bfeb5ac5/stl/inc/algorithm#L6384-L6406
It's a very common implementation. 
And what you posted is an exact search, not lower_bound needed here.

File: storage/diemdb/src/event_store/mod.rs (hunk 2 above)
Comment (msmouse) [reply storage/diemdb/src/event_store/mod.rs:300]: Just to talking in the C++ half open range semantics that I'm familiar with.

**Review by kchalkias — APPROVED** [review]
LGTM

File: storage/diemdb/src/event_store/mod.rs [hunk 4]
```diff
@@ -238,6 +282,84 @@ impl EventStore {
...
+        &self,
+        event_key: &EventKey,
+        mut comp: C,
+        ledger_version: Version,
+    ) -> Result<Option<u64>>
+    where
+        C: FnMut(&ContractEvent) -> Result<bool>,
+    {
+        let mut begin = 0u64;
+        let mut end = match self.get_latest_sequence_number(ledger_version, event_key)? {
```
Comment (davidiw) [inline storage/diemdb/src/event_store/mod.rs:298]: why you no like map_or_else

File: storage/diemdb/src/event_store/mod.rs [hunk 5]
```diff
@@ -238,6 +282,84 @@ impl EventStore {
...
+                let mid = begin + step;
+                let event = self.get_event_by_key(event_key, mid, ledger_version)?;
+                if comp(&event)? {
+                    begin = mid + 1;
+                    count -= step + 1;
+                } else {
+                    count = step;
+                }
+            }
+        }
```
Comment (davidiw) [inline storage/diemdb/src/event_store/mod.rs:320]: oh my... coding interviews...

aren't we supposed to break if we find it? shouldn't this be a > go up, < go down, == stop.

File: storage/diemdb/src/event_store/mod.rs (hunk 4 above)
Comment (msmouse) [reply storage/diemdb/src/event_store/mod.rs:298]: The None branch quits this method, `map_or_else()` doesn't do that?

File: storage/diemdb/src/event_store/mod.rs (hunk 5 above)
Comment (msmouse) [reply storage/diemdb/src/event_store/mod.rs:320]: Yeah.. you need to distinguish `lower_bound`, `upper_bound`, `equal_range` and the toy exact binary search found on the internet that returns a random index among duplications. Great preparation to do before a C++ interview (I made that up).

**Pull request merged**